server/src/main/java/org/elasticsearch/gateway/GatewayService.java

```diff
@@ -141,77 +147,92 @@ public void clusterChanged(final ClusterChangedEvent event) {
             logger.debug(
                 "not recovering from gateway, nodes_size (data) [{}] < recover_after_data_nodes [{}]",
                 nodes.getDataNodes().size(),
                 recoverAfterDataNodes
             );
-        } else {
-            boolean enforceRecoverAfterTime;
-            String reason;
-            if (expectedDataNodes == -1) {
-                // no expected is set, honor recover_after_data_nodes
-                enforceRecoverAfterTime = true;
-                reason = "recover_after_time was set to [" + recoverAfterTime + "]";
-            } else if (expectedDataNodes <= nodes.getDataNodes().size()) {
-                // expected is set and satisfied so recover immediately
-                enforceRecoverAfterTime = false;
-                reason = "";
+            return;
+        }
+
+        // At this point, we know this node is qualified for state recovery
+        // But we still need to check whether a previous one is running already
+        final SubscribableListener<Void> thisRecoveryPlanned;
+        synchronized (this) {
+            if (recoveryPlanned == null) {
+                recoveryPlanned = thisRecoveryPlanned = new SubscribableListener<>();
             } else {
-                // expected is set but not satisfied so wait until it is satisfied or times out
-                enforceRecoverAfterTime = true;
-                reason = "expecting [" + expectedDataNodes + "] data nodes, but only have [" + nodes.getDataNodes().size() + "]";
+                thisRecoveryPlanned = null;
             }
-            performStateRecovery(enforceRecoverAfterTime, reason);
         }
-    }
 
-    private void performStateRecovery(final boolean enforceRecoverAfterTime, final String reason) {
-        if (enforceRecoverAfterTime && recoverAfterTime != null) {
-            if (scheduledRecovery.compareAndSet(false, true)) {
-                logger.info("delaying initial state recovery for [{}]. {}", recoverAfterTime, reason);
-                threadPool.schedule(new AbstractRunnable() {
-                    @Override
-                    public void onFailure(Exception e) {
-                        logger.warn("delayed state recovery failed", e);
-                        resetRecoveredFlags();
-                    }
-
-                    @Override
-                    protected void doRun() {
-                        if (recoveryInProgress.compareAndSet(false, true)) {
-                            logger.info("recover_after_time [{}] elapsed. performing state recovery...", recoverAfterTime);
-                            runRecovery();
-                        }
-                    }
-                }, recoverAfterTime, threadPool.generic());
+        if (thisRecoveryPlanned == null) {
+            logger.debug("state recovery is in progress");
+            return;
+        }
+
+        // This node is ready to schedule state recovery
+        thisRecoveryPlanned.addListener(new ActionListener<>() {
```

Comment:
I wanted to use `andThen` but it does not expose exception to the consumer while we need the exception here to handle timeout.